### PR TITLE
UI/update app source

### PR DESCRIPTION
**Pull request by scures**
### Summary
Fixes # https://github.com/epinio/ui/issues/171

Allows the user to edit an application and redeploy from another source

### How to test:
- Go to https://localhost:8005/epinio/c/default/applications
- Click on one of your application and check the Config tab (top right), the applications source tab **shouldn't** be displayed!
- Click on the Actions button (next to Config, 3 dots) and click Edit Config, now the Source tab should show with the wizard.
- Choose a different source/application and follow the steps, everything should stay the same (envs, configs,...) but the app source should be different.
-

### CODE REVIEW

File: pkg/epinio/edit/applications.vue [hunk 1]
```diff
@@ -100,15 +142,59 @@ export default Vue.extend<Data, any, any, any>({
...
+      const { appChart, ...cleanChanges } = changes;
+
+      this.value.configuration = this.value.configuration || {};
+
+      if (appChart) {
+        // app chart actually belongs in config, so stick it in there
+        this.set(this.value.configuration, { appchart: appChart });
+      }
+
+      if (changes.type === APPLICATION_SOURCE_TYPE.GIT_HUB) {
```
Comment (richard-cox) [inline pkg/epinio/edit/applications.vue:169]: Shouldn't this be doing the stripping of env vars if the user switches back out of a app env var based app source? If so do we need to do it in the app model?

File: pkg/epinio/edit/applications.vue [hunk 2]
```diff
@@ -100,15 +142,59 @@ export default Vue.extend<Data, any, any, any>({
...
+        this.value.configuration.environment = this.value.configuration.environment || {};
+        const githubEnvVar: EPINIO_APP_ENV_VAR_GITHUB = {
```
Comment (richard-cox) [inline pkg/epinio/edit/applications.vue:171]: Creating the env var object needs to stay in sync with the other  place this is done in app create, probably needs a common function

File: pkg/epinio/components/application/AppProgressUpdateSource.vue [hunk 3]
```diff
@@ -0,0 +1,261 @@
+<script lang="ts">
```
Comment (richard-cox) [inline pkg/epinio/components/application/AppProgressUpdateSource.vue:1]: There's huge C&P between this and `AppProgress`. The mode (create / edit) can be passed to `AppProgress` instead and the steps can be tweaked given that param

File: pkg/epinio/l10n/en-us.yaml [hunk 4]
```diff
@@ -218,6 +218,9 @@ epinio:
       bind_configurations:
         label: Bind Configurations
         description: Configurations will be bound to the new Application
+      updateSource:
+        label: Updated
```
Comment (richard-cox) [inline pkg/epinio/l10n/en-us.yaml:222]: Updating

**Review by richard-cox — CHANGES_REQUESTED** [review]
- The starting value for the edit source wizard source step should be the current source (seen with richard / demo-app application with source of container image)
- The final step before progress in the wizard in the create flow is different from previous buttons (Blue Next --> Yellow Create). The same needs to happen in the edit source wizard to show the user that clicking the button will apply their changes

File: pkg/epinio/pages/c/_cluster/applications/index.vue [hunk 5]
```diff
@@ -60,7 +58,6 @@ export default {
 
 <template>
   <Loading v-if="$fetchState.pending" />
-  <EpinioIntro v-else-if="!hasNamespaces" />
```
[the remark this answers is not in the document]
Comment (richard-cox) [reply pkg/epinio/pages/c/_cluster/applications/index.vue:63]: Actually, what is it?  Is the content now on the home page? if so the component can be removed

**Review by richard-cox — CHANGES_REQUESTED** [review]
- Might be wrong here, but the app details / deployment details should show origin as github for github apps right? in the app `richard2 / github` its `Git` and doesn't show the branch.
- The initial values of step 1 when editing source are still the default ones, rather than the actual app values

**Review by richard-cox — CHANGES_REQUESTED** [review]
There's an issue here from https://github.com/rancher/dashboard/pull/6469/files#diff-da9aa82757cf37fb40c262464640de834259f5e76621b251c425f657e323af22R164. 

application-action createDeployOrigin supplies a `kind`, which is a APPLICATION_MANIFEST_SOURCE_TYPE. The comment for that references what the backend things is valid (none: 0, path: 1, git: 2, container: 3). our github/gitlab is their git, they don't differentiate.

Steps to resolve
1. Remove `GIT_BUG` from APPLICATION_MANIFEST_SOURCE_TYPE
2. createDeployOrigin should set APPLICATION_MANIFEST_SOURCE_TYPE.GIT for github 
3. createDeployOrigin should set a `type` of APPLICATION_SOURCE_TYPE.GIT_HUB within our env var
4. add a `get` in the `application` model `type` that returns the manifest kind unless we have a type in the env var
5. everywhere that needs to know the type uses the source (and in the create/edit flow the initial value comes from there to set in the params object that contains the output of the wizard)

@torchiaf you're hitting similar issues in the gitlab PR. we should get this one in and yours should be straight forward to address

**Comment by richard-cox** [comment]
I've made a lot of changes and fixed a lot of things. More detail in the commit message. Briefly
- Updated the app env var to include more information
- The additional information is shown in the app detail deployment page as part of the existing way we show app source info
- The additional information supports editing an applications source (including builder image, but not app chart)
- Fixed a lot of things around the app source component and accepting existing values

I've also sneaked in the Application Template --> Application Chart change

@torchiaf I've changed a lot, so I don't think I'm in a good position to review this. Can you take a look?

@mmartin24 Lots has changed in the way we create/deploy an application, persist information about that deployment and use it all in the new edit application source feature. It might break some of your e2e tests, but definitely needs some validating with a keen eye. I've listed some things here, but it's not an exhaustive list. Let me know if you'd like a call to run through these.

Some test cases to cover might include
- Deploy applications for each source type. View each applications detail page
- For an application from each source type, edit the application's source in a trivial way (without changing type)
- For an application from a small range of source types, edit the applicatin's source type in a major way (change the source type)
- Create an application using github as the source. Edit the application and remove the application's env vars. View details for the application (should not error)
- Ensure that an applications builder image can be changed (for example to paketobuildpacks/builder:base). Can be done in conjunction with above. Only applicable to some source types

**Comment by torchiaf** [comment]
> I've made a lot of changes and fixed a lot of things. More detail in the commit message. Briefly
> 
> * Updated the app env var to include more information
> * The additional information is shown in the app detail deployment page as part of the existing way we show app source info
> * The additional information supports editing an applications source (including builder image, but not app chart)
> * Fixed a lot of things around the app source component and accepting existing values
> 
> I've also sneaked in the Application Template --> Application Chart change
> 
> @torchiaf I've changed a lot, so I don't think I'm in a good position to review this. Can you take a look? 

Working on it :+1:

**Review by torchiaf — CHANGES_REQUESTED** [review]
It works fine, tested every kind of App Source.
There is a crash when the application's name is clicked in the Application table but it's trivial. I also left some comments about optional refactoring.

File: pkg/epinio/models/applications.js [hunk 6]
```diff
@@ -237,62 +236,191 @@ export default class EpinioApplicationModel extends EpinioNamespacedResource {
...
+    if (!this.origin && !appEnvVarSource) {
       return undefined;
     }
     const appChart = {
       label: 'App Chart',
       value: this.configuration.appchart
     };
+    const builderImage = {
+      label: 'Builder Image',
+      value: appEnvVarSource.builderImage
```
[the remark this answers is not in the document]
Comment (torchiaf) [reply pkg/epinio/models/applications.js:365]: done

**Review by torchiaf — APPROVED** [review]
I a look to the last commits @richard-cox . LGTM

**Comment by mmartin24** [comment]
> I've made a lot of changes and fixed a lot of things. More detail in the commit message. Briefly
> 
> * Updated the app env var to include more information
> * The additional information is shown in the app detail deployment page as part of the existing way we show app source info
> * The additional information supports editing an applications source (including builder image, but not app chart)
> * Fixed a lot of things around the app source component and accepting existing values
> 
> I've also sneaked in the Application Template --> Application Chart change
> 
> @torchiaf I've changed a lot, so I don't think I'm in a good position to review this. Can you take a look? Edit - Your gitlab PR will need updating after this, but hopefully should just mean searching for the GITHUB app source type in a similar way you updated things before.
> 
> @mmartin24 Lots has changed in the way we create/deploy an application, persist information about that deployment and use it all in the new edit application source feature. It might break some of your e2e tests, but definitely needs some validating with a keen eye. I've listed some things here, but it's not an exhaustive list. Let me know if you'd like a call to run through these.
> 
> Some test cases to cover might include
> 
> * Deploy applications for each source type. View each applications detail page
> * For an application from each source type, edit the application's source in a trivial way (without changing type)
> * For an application from a small range of source types, edit the applicatin's source type in a major way (change the source type)
> * Create an application using github as the source. Edit the application and remove the application's env vars. View details for the application (should not error)
> * Ensure that an applications builder image can be changed (for example to paketobuildpacks/builder:base). Can be done in conjunction with above. Only applicable to some source types

@richard-cox , I started having a look on the main ticket and I am not sure if we have a bug (commented [here](https://github.com/epinio/ui/issues/171#issuecomment-1508040861)) or I am not configuring properly, so I'd like please to take your call and meet for a small demo. Thanks

**Pull request merged**